portfolio/src/main/java/com/google/sps/servlets/FileHandlerServlet.java

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import com.google.appengine.repackaged.com.google.gson.Gson;
+import com.google.appengine.api.datastore.Query;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/files")
+public class FileHandlerServlet extends HttpServlet {
+
+private static final String CONTENT_TYPE_JSON = "application/json;";
```

Comment:
let's break down your question about constants a little more. generally, i'm a bit suspicious of files like a hypothetical Constants.java. they end up serving as dumping grounds for pretty much every constant, and the reason that can be a bad thing is that they will all have the same visibility. pretty much every piece of code, whether a constant or whatever, you want to have the minimum visibility possible. to consider some examples:

PATH_HOME_PAGE - this seems like it should have pretty high visibility, any piece of code that wants to redirect to the home page or something needs this constant

MAX_COMMENTS - there should be no reason for anyone outside of the code dealing with comments to see this. if there is outside code that needs to see this it's a good indication that code is doing something weird or incorrect, and may need to be fixed.

KEY_IMAGE_BLOBKEY - same logic, only code dealing with getting/setting images should ever need to see this.

Immediately we can see that we might require at least 3 different visibilities here. Thus, I would say the simplest solution is to keep each in the file where they are needed. PATH_HOME_PAGE etc, could potentially be extracted out to some PathConstants.java, etc, if you so desired.

One last comment re CONTENT_TYPE_JSON. This is a slightly unusual constant, in that its value can never ever change. It's defined in god knows how many specs, and if it's value ever changed, half the internet would break. Constants like these are more acceptable to use directly as literals in code, since we know there is a 0% chance they could change. Constants like KEY_IMAGE_URL, we might want to change in the future, and it makes sense to have a central reference.

A final thought on duplication: It's easier to consolidate duplication than to unconsolidate something. If you're in doubt, it's usually better to err on the side of duplicating (constants for example). If you change your mind, it's easy to centralize. If you've already centralized however, and later decide you want to change that, you may find it harder.